h5netcdf/core.py

```diff
@@ -1192,9 +1277,17 @@ def __init__(self, path, mode="r", invalid_netcdf=False, phony_dims=None, **kwar
                 )
                 raise TypeError(msg)
             self.decode_vlen_strings = True
+            # default _auto_complex to True for legacyapi
+            # this creates files which are readable by netCDF4 with nc-complex
+            # as well as h5netcdf
+            if self._auto_complex is None:
+                self._auto_complex = True
         else:
             if self.decode_vlen_strings is None:
                 self.decode_vlen_strings = False
+            # default _auto_complex to False for new API
+            if self._auto_complex is None:
+                self._auto_complex = False
```

Comment:
This mismatch between the legacy and new APIs is a little strange and could surprise users. A more conservative choice (which could be changed later) is to defautl to `auto_complex=False` and require users to explicitly opt-in to the new behavior.

Are arrays written with `auto_complex=True` automatically read as complex valued with h5py or old versions of h5netcdf? If not, I would stick with `auto_complex=False` as a default in both cases until a concensus emerges that this is the right way to handle complex values in netCDF files.

Reply:
The only difference for h5py/h5netcdf is the additional committed (named) type within the file on write. Old versions can read such files, relying on the type retrieved from the dataset. New versions of h5netcdf also gather the type in `.cmptypes`.

But going conservative first is a good idea anyway. netcdf4-python does the same, @ZedThree, right?

Thanks Stefan for the review, much appreciated.  😊

Reply:
I've checked again, see https://github.com/h5netcdf/h5netcdf/pull/227#issuecomment-1902243420